active/0000-registry-reform.md

```diff
@@ -0,0 +1,109 @@
+rather than `_registry`, as it can be assumed that only framework developers
+will reference this property.
+
+# Detailed design
+
+`Application` will expose the following methods from its internally maintained
+registry:
+
+* `register`
+* `inject`
+* `registerOptions` - mapped to `Registry#options`
+* `registerOptionsForType` - mapped to `Registry#optionsForType`
+
+`ApplicationInstance` will also expose the the same methods. However, these
+methods will be exposed from its own internally maintained registry, which
+has the associated `Application`'s registry configured as a "fall back". No
+direct path will be provided from the `ApplicationInstance` to the
+`Application`'s registry.
+
+`ApplicationInstance` will also expose the following methods from its
+internally maintained container:
+
+* `lookup`
+* `lookupFactory`
+
+`ApplicationInstance` will cease exposing `container`, `registry`, and
+`applicationRegistry` publicly.
+
+`Application` initializers will receive a single argument to `initialize`:
+`application`.
+
+Likewise, `ApplicationInstance` initializers will receive a single argument
+to `initialize`: `applicationInstance`.
+
+`Container` and `Registry` will be made fully private and documented as
+such. Each `Container` will freely reference its associated `Registry` as
+`registry` rather than `_registry`.
+
+[ember-test-helpers](https://github.com/switchfly/ember-test-helpers)
+will provide an `isolatedApplicationInstance` method instead of an
+`isolatedContainer` for unit testing. A mechanism will be developed to specify
+which initializers should be engaged in the initialization of this instance.
+In this way, we can avoid duplication of registration logic, as is currently
+done in a most un-DRY manner in the [isolatedContainer](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/isolated-container.js#L56-L79).
+
+# Drawbacks
+
+This refactor will require maintaining backwards compatibility and
+deprecation warnings until Ember 2.0. This will temporarily increase
+internal code complexity and file sizes.
+
+# Alternatives
+
+The obvious alternative is to make `Container` and `Registry` fully public
+and documented. An application's registry would be available as a `registry`
+property. An application instance's container would remain available as
+`container`.
+
+We could still pass an `Application` into application initializers
+and an `ApplicationInstance` into application instance initializers.
+
+If this alternative is taken, I would suggest that `Application` should
+deprecate `register` and `inject` in favor of calling the equivalents on its
+public `registry`.
+
+Regardless of which alternative is chosen, we should ensure that the public
+aspects of container and registry usage are well documented.
+
+# Unresolved questions
+
+* Are the public methods listed above sufficient or should any others be
+exposed?
+
+* What mechanism should be used to engage initializers in unit and
```

Comment:
All of Ember's "unit" tests rely on the concept of a container that is initialized. The [isolatedContainer](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/isolated-container.js) is not truly "isolated", but relies on registrations performed by Ember and even Ember Data. I'm sure we can all agree that the code in `isolatedContainer` is un-DRY and a temporary hack, so let's solve this problem more elegantly.

As [the discussion about integration vs. unit testing components](https://github.com/switchfly/ember-test-helpers/pull/38) makes clear, we need a broader discussion about dependency injection, what's exposed to tests, and ember-cli's conventions regarding integration and unit tests. This goes beyond the scope of this RFC, but definitely overlaps it.

@stefanpenner I'll ping you soon to try to set up that overdue hangout / discussion.